source/common/upstream/thread_aware_lb_impl.cc

```diff
@@ -185,5 +185,121 @@ LoadBalancerPtr ThreadAwareLoadBalancerBase::LoadBalancerFactoryImpl::create() {
+  }
+  return static_cast<double>(host.stats().rq_active_.value()) / slots;
+}
+
+HostConstSharedPtr
+ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer::chooseHost(uint64_t hash,
+                                                                        uint32_t attempt) const {
+
+  // This is implemented based on the method described in the paper
+  // https://arxiv.org/abs/1608.01350. For the specified `hash_balance_factor`, requests to any
+  // upstream host are capped at `hash_balance_factor/100` times the average number of requests
+  // across the cluster. When a request arrives for an upstream host that is currently serving at
+  // its max capacity, linear probing is used to identify an eligible host. Further, the linear
+  // probe is implemented using a random jump on hosts ring/table to identify the eligible host
+  // (this technique is as described in the paper https://arxiv.org/abs/1908.08762 - the random jump
+  // avoids the cascading overflow effect when choosing the next host on the ring/table).
+  //
+  // If weights are specified on the hosts, they are respected.
+  //
+  // This is an O(N) algorithm, unlike other load balancers. Using a lower `hash_balance_factor`
+  // results in more hosts being probed, so use a higher value if you require better performance.
+
+  if (normalized_host_weights_.empty()) {
+    return nullptr;
+  }
+
+  HostConstSharedPtr host = hashing_lb_ptr_->chooseHost(hash, attempt);
+  if (host == nullptr) {
+    return nullptr;
+  }
+  const double weight = normalized_host_weights_map_.at(host);
+  double overload_factor = hostOverloadFactor(*host, weight);
+  if (overload_factor <= 1.0) {
+    ENVOY_LOG_MISC(debug,
+                   "ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer::chooseHost: "
+                   "selected host #{} (attempt:1)",
+                   host->address()->asString());
+    return host;
+  }
+
+  // When a host is overloaded, we choose the next host in a random manner rather than picking the
+  // next one in the ring. The random sequence is seeded by the hash, so the same input gets the
+  // same sequence of hosts all the time.
+  const uint32_t num_hosts = normalized_host_weights_.size();
+  auto host_index = std::vector<uint32_t>(num_hosts);
+  for (uint32_t i = 0; i < num_hosts; i++) {
+    host_index[i] = i;
+  }
+
+  // Not using Random::RandomGenerator as it does not take a seed. Seeded RNG is a requirement
+  // here as we need the same shuffle sequence for the same hash every time.
+  // Further, not using std::default_random_engine and std::uniform_int_distribution as they
+  // are not consistent across Linux and Windows platforms.
+  const uint64_t seed = hash;
+  std::mt19937 random(seed);
+
+  // generates a random number in the range [0,k) uniformly.
+  auto uniform_int = [](std::mt19937& random, uint32_t k) -> uint32_t {
+    uint32_t x = k;
+    while (x >= k) {
+      x = random() / ((static_cast<uint64_t>(random.max()) + 1u) / k);
```

Comment:
If the return type of random.max() is uint64_t, then adding +1 results in an integer overflow error and divide by 0.

Consider removing the +1.

Reply:
The default word size of std::mt19937 is 32. The casting to uint64_t was precisely to avoid the overflow when +1'd.